util/progress/progressui/display.go

```diff
@@ -159,6 +162,42 @@ func (v *vertex) update(c int) {
 	v.count += c
 }
 
+type vertexGroup struct {
+	*vertex
+	subVtxs map[digest.Digest]client.Vertex
+}
+
+func (vg *vertexGroup) mergeSubVtxs() *client.Vertex {
+	newVtx := *vg.Vertex
+	newVtx.Completed = &time.Time{}
+	newVtx.Cached = true
+	for _, subVtx := range vg.subVtxs {
+		// Group start time is the earliest start time of the vtxs (or nil if none have started)
+		if newVtx.Started == nil || (subVtx.Started != nil && subVtx.Started.Before(*newVtx.Started)) {
+			newVtx.Started = subVtx.Started
+		}
+
+		// Group is considered not completed if any vtxs are not completed. If all vtxs
```

Comment:
The duration if the group is not `max(completed)-min(started)`. We need to sum the times when the group goes to completed and add the current time since it last went to started. We should do the same for non-groups as well when they go to started state multiple times. Possibly tracking both cases with the same object.

Reply:
> We need to sum the times when the group goes to completed and add the current time since it last went to started.

I'm not following. Are you saying the total time for a vertex group should be the sum of the intervals each group member was running for? If so, I feel like that could be extremely confusing. Say there's 5 file-ops in a progress group that each take 2 seconds, but the file ops all run in parallel. The user would see that their progress group took 10 seconds but in reality it only took 2 seconds. This might make sense if they could see the sub vertexes in the group (and that their subtimes add to 10), but by design they can't see that.

Reply:
Are you saying that we should sum the start/stop times but account for overlap? I.e. if the start stop intervals of the group members are
1. (0, 2)
1. (1, 2)
1. (3, 4)

Then the total time should be displayed as 3? If so, that makes more sense to me.

Reply:
Not total of everything but it should only be active time.

Eg. common case: copy to scratch + merge on top of base. Copy to scratch will run instantly on the beginning of the build because it has no dependencies. Lets say it takes 0.5 seconds. Then base is built, maybe takes 5 min. Then merge can run with another 0.5s. The total time for the progressgroup should be 1s. Not 5:01.

When 2 refs are active at the same time then we do not need to (and shouldn't) double count. If vertex1 starts at 0:00 and ends in 0:10 and vertex2 in 0:05 and ends in 0:15 then total time is 15sec.

Same case is for the non-groups. If the vertex is started for cachemap and later for exec then progress should show the sum. This happens on image pull for example and is confusing atm.

Reply:
Yes. Same as your last comment. I didn't see it before.

Reply:
Awesome, makes sense. I am still working on this update as it's a bit more involved, I'll ping once I push it.

Reply:
@tonistiigi updated with an extra commit implementing this (works with both normal vertexes and progress groups). I tested locally by adding temporary `time.Sleep` statements to various places to see the vertexes starting/stopping. It seems to work by my eye but it's not something easy to automatically test so let me know if anything seems off.